**Quiz 5 & 6**


In [43]:
import pandas as pd
import warnings
import numpy as np
from scipy.stats import norm
warnings.filterwarnings("ignore")

In [44]:
from google.colab import drive
drive.mount('/content/drive')
import pandas as pd
data = pd.read_csv('/content/drive/My Drive/Colab Notebooks/Earthquakes_database.csv')
data.head()

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Date      Time  Latitude  Longitude        Type  Depth  Depth Error  \
0  1/2/1965  13:44:18    19.246    145.616  Earthquake  131.6          NaN   
1  1/4/1965  11:29:49     1.863    127.352  Earthquake   80.0          NaN   
2  1/5/1965  18:05:58   -20.579   -173.972  Earthquake   20.0          NaN   
3  1/8/1965  18:49:43   -59.076    -23.557  Earthquake   15.0          NaN   
4  1/9/1965  13:32:50    11.938    126.427  Earthquake   15.0          NaN   

   Depth Seismic Stations  Magnitude Magnitude Type  ...  \
0                     NaN        6.0             MW  ...   
1                     NaN        5.8             MW  ...   
2                     NaN        6.2             MW  ...   
3                     NaN        5.8             MW  ...   
4                     NaN        5.8             MW  ...   

   Magnitude Seismic Stations  Azimuthal Gap  Horizontal Distance  \
0                         NaN            NaN                  NaN   
1                         NaN            NaN                  NaN   
2                         NaN            NaN                  NaN   
3                         NaN            NaN                  NaN   
4                         NaN            NaN                  NaN   

   Horizontal Error  Root Mean Square            ID  Source Location Source  \
0               NaN               NaN  ISCGEM860706  ISCGEM          ISCGEM   
1               NaN               NaN  ISCGEM860737  ISCGEM          ISCGEM   
2               NaN               NaN  ISCGEM860762  ISCGEM          ISCGEM   
3               NaN               NaN  ISCGEM860856  ISCGEM          ISCGEM   
4               NaN               NaN  ISCGEM860890  ISCGEM          ISCGEM   

  Magnitude Source     Status  
0           ISCGEM  Automatic  
1           ISCGEM  Automatic  
2           ISCGEM  Automatic  
3           ISCGEM  Automatic  
4           ISCGEM  Automatic  

[5 rows x 21 columns]

# მონაცემების მსგავსი კლასებად დაგხვემარება შემდგომ რომ კლასიფიკაციის მოდელი მარტივად ავაწყოთ.

In [45]:
data['magnitude_classes'] = pd.cut(data['Magnitude'],
                                    bins=[3, 3.9, 4.9, 5.9, 6.9, 7.9, float('inf')],
                                    labels=['minor', 'light', 'moderate', 'strong', 'major', 'great'])

In [46]:
data

Date      Time  Latitude  Longitude        Type   Depth  \
0        1/2/1965  13:44:18   19.2460   145.6160  Earthquake  131.60   
1        1/4/1965  11:29:49    1.8630   127.3520  Earthquake   80.00   
2        1/5/1965  18:05:58  -20.5790  -173.9720  Earthquake   20.00   
3        1/8/1965  18:49:43  -59.0760   -23.5570  Earthquake   15.00   
4        1/9/1965  13:32:50   11.9380   126.4270  Earthquake   15.00   
...           ...       ...       ...        ...         ...     ...   
23407  12/28/2016   8:22:12   38.3917  -118.8941  Earthquake   12.30   
23408  12/28/2016   9:13:47   38.3777  -118.8957  Earthquake    8.80   
23409  12/28/2016  12:38:51   36.9179   140.4262  Earthquake   10.00   
23410  12/29/2016  22:30:19   -9.0283   118.6639  Earthquake   79.00   
23411  12/30/2016  20:08:28   37.3973   141.4103  Earthquake   11.94   

       Depth Error  Depth Seismic Stations  Magnitude Magnitude Type  ...  \
0              NaN                     NaN        6.0             MW  ...   
1              NaN                     NaN        5.8             MW  ...   
2              NaN                     NaN        6.2             MW  ...   
3              NaN                     NaN        5.8             MW  ...   
4              NaN                     NaN        5.8             MW  ...   
...            ...                     ...        ...            ...  ...   
23407          1.2                    40.0        5.6             ML  ...   
23408          2.0                    33.0        5.5             ML  ...   
23409          1.8                     NaN        5.9            MWW  ...   
23410          1.8                     NaN        6.3            MWW  ...   
23411          2.2                     NaN        5.5             MB  ...   

       Azimuthal Gap  Horizontal Distance  Horizontal Error  Root Mean Square  \
0                NaN                  NaN               NaN               NaN   
1                NaN                  NaN               NaN               NaN   
2                NaN                  NaN               NaN               NaN   
3                NaN                  NaN               NaN               NaN   
4                NaN                  NaN               NaN               NaN   
...              ...                  ...               ...               ...   
23407          42.47                0.120               NaN            0.1898   
23408          48.58                0.129               NaN            0.2187   
23409          91.00                0.992               4.8            1.5200   
23410          26.00                3.553               6.0            1.4300   
23411          97.00                0.681               4.5            0.9100   

                 ID  Source Location Source Magnitude Source     Status  \
0      ISCGEM860706  ISCGEM          ISCGEM           ISCGEM  Automatic   
1      ISCGEM860737  ISCGEM          ISCGEM           ISCGEM  Automatic   
2      ISCGEM860762  ISCGEM          ISCGEM           ISCGEM  Automatic   
3      ISCGEM860856  ISCGEM          ISCGEM           ISCGEM  Automatic   
4      ISCGEM860890  ISCGEM          ISCGEM           ISCGEM  Automatic   
...             ...     ...             ...              ...        ...   
23407    NN00570710      NN              NN               NN   Reviewed   
23408    NN00570744      NN              NN               NN   Reviewed   
23409    US10007NAF      US              US               US   Reviewed   
23410    US10007NL0      US              US               US   Reviewed   
23411    US10007NTD      US              US               US   Reviewed   

      magnitude_classes  
0                strong  
1              moderate  
2                strong  
3              moderate  
4              moderate  
...                 ...  
23407          moderate  
23408          moderate  
23409          moderate  
23410            strong  
23411          moderate  

[23412 rows x 22 columns]

ხაზოვანი რეგრესიის მოდელი მომზადებულია გეოგრაფიულ კოორდინატებს (გრძედისა და განედის) და მიწისძვრების სიდიდეს შორის კავშირის შესასწავლად.


In [51]:

from sklearn.linear_model import LinearRegression

from sklearn.impute import SimpleImputer

features = ['Latitude', 'Longitude', ]
target = 'Magnitude'


X = data[features]
y = data[target]


imputer = SimpleImputer(strategy='mean')
X = pd.DataFrame(imputer.fit_transform(X), columns=X.columns)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


model = LinearRegression()

model.fit(X_train, y_train)


y_pred = model.predict(X_test)


mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f'Mean Squared Error: {mse}')
print(f'R-squared: {r2}')

print('Coefficients:', model.coef_)
print('Intercept:', model.intercept_)


Mean Squared Error: 0.18475232430091937
R-squared: -0.0016566730482006076
Coefficients: [0.00051777 0.00011727]
Intercept: 5.875776985856692


მაგნიტუდის მისახებად გამოყენებული იყო ფორმულა :
Magnitude=5.88+(0.00052×Latitude)+(0.00012×Longitude)

მიუხედავად იმისა რომ MSE (კვადრატული სხვაობა პროგნოზირებულ და რეალურ სიდიდეებს შორის) პატარა არის ,  ვხედავთ რომ R-squared უარყოფით მნიშვნელობას ღებულობს,რაც იმაზე მიგვითითებს რომ მოდელი კარგად არ მუშაობს მაგნიტუდის დისპერსიის ასახსნელად


In [48]:
from sklearn.tree import DecisionTreeRegressor
import matplotlib.pyplot as plt


features = ['Latitude', 'Longitude', 'Depth']
target = 'Magnitude'

X = data[features]
y = data[target]


X = X.fillna(X.mean())

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

model = DecisionTreeRegressor(random_state=42)

model.fit(X_train, y_train)

y_pred = model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f'Mean Squared Error: {mse}')
print(f'R-squared: {r2}')




Mean Squared Error: 0.34830390775144143
R-squared: -0.8883710111259477


MSE-სა და   R_SQUARED-ს შორის მსგავის დამოკიდებულება მიგვითითებს  overfitting-ზე ან ისევ მოდელის არასწორობაზე


In [49]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

features = ['Latitude', 'Longitude', 'Depth']
target = 'magnitude_classes'

X = data[features]
y = data[target]


X = X.fillna(X.mean())

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

model = LogisticRegression(random_state=42)

model.fit(X_train, y_train)

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
classification_rep = classification_report(y_test, y_pred)

print(f'Accuracy: {accuracy}')
print('Confusion Matrix:\n', conf_matrix)
print('Classification Report:\n', classification_rep)

Accuracy: 0.6843903480674781
Confusion Matrix:
 [[   0    0    9    0]
 [   0    0  143    0]
 [   0    0 3203    1]
 [   0    0 1325    2]]
Classification Report:
               precision    recall  f1-score   support

       great       0.00      0.00      0.00         9
       major       0.00      0.00      0.00       143
    moderate       0.68      1.00      0.81      3204
      strong       0.67      0.00      0.00      1327

    accuracy                           0.68      4683
   macro avg       0.34      0.25      0.20      4683
weighted avg       0.66      0.68      0.56      4683



In [50]:
from sklearn.tree import DecisionTreeClassifier
features = ['Latitude', 'Longitude', 'Depth', 'Depth Error', 'Horizontal Distance']
target = 'magnitude_classes'

X = data[features]
y = data[target]


X = X.fillna(X.mean())

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

model = DecisionTreeClassifier(random_state=42)


model.fit(X_train, y_train)


y_pred = model.predict(X_test)


accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
classification_rep = classification_report(y_test, y_pred)

print(f'Accuracy: {accuracy}')
print('Confusion Matrix:\n', conf_matrix)
print('Classification Report:\n', classification_rep)

Accuracy: 0.5846679478966474
Confusion Matrix:
 [[   0    0    6    3]
 [   1    5   82   55]
 [   4   93 2271  836]
 [   1   50  814  462]]
Classification Report:
               precision    recall  f1-score   support

       great       0.00      0.00      0.00         9
       major       0.03      0.03      0.03       143
    moderate       0.72      0.71      0.71      3204
      strong       0.34      0.35      0.34      1327

    accuracy                           0.58      4683
   macro avg       0.27      0.27      0.27      4683
weighted avg       0.59      0.58      0.59      4683

